README.md
# monopole_MJD

1. This package was modify from the $GEANT4DIR/examples/extended/exoticphysics/monopole/.

2. Output the stepping parameters.

3. Use Majorana Demonstrator detectors (defined in Detectorposition.txt).

4. Assume the monopoles randomly injecting from the hemisphere 

## Compile monopole_MJD
1. $mkdir ./monopole-build/

2. $cd ./monopole-build/

3. $cmake -DGeant4_DIR=/mjsw/GEANT/install/lib64/Geant4-10.4.3/ ../

4. $make

## Execute monopole_MJD

1. Need copy ./Detectorposition.txt to ./monopole-build/

2. $./monopole monopole.in
This generates monopole.root

3. Need to modify the monopole mass in ./src/G4MonopolePhysics.cc and re-compile before generating Monte Carlo
Modify the line "fMonopoleMass = 1e12*GeV;"

## Execute Scripts
1. $./analysis/run.pl 1e12
1e12 is the monopole mass used in G4MonopolePhysics.cc
The script will run replace.pl and generate Monte Carlo with the energy from 1e-5 GeV to 1e10 GeV.

2. $./analysis/replace.pl $erg
$erg is the monopole energy. The script will modify the monopole energy in monopole.in.
./analysis/replacemass.pl $mass
$mass is the monopole mass.

## Analysis Scripts
1. $./analysis/analysis.cc 
   $./analysis/analysis.pl
This script reconstructed and analyzed the raw data (monopole.root)

## Monte Carlo Units
The unit of Monte Carlo: time(ns), energy(MeV), length(mm)


History
$Id$
-------------------------------------------------------------------

     =========================================================
     Geant4 - an Object-Oriented Toolkit for Simulation in HEP
     =========================================================

                    Monopole Example History file
                    -----------------------------
This file should be used by the G4 example coordinator to briefly
summarize all major modifications introduced in the code and keep
track of all tags.

     ----------------------------------------------------------
     * Reverse chronological order (last date on top), please *
     ----------------------------------------------------------

21 November 17: V.Ivantchenko (monopole-V10-03-04)
- Simplified histograms defition and filling

20 November 17: V.Ivantchenko (monopole-V10-03-03)
- Fixed lost functionality of histograms and monopole parameter
  definitions

26 October 17: D. Sorokin (monopole-V10-03-01)
- G4MonopoleTransportation.cc - build fix

23 June 17: S. Elles (monopole-V10-03-00)
- add MT interface
- switch ActionInitialization, Run and RunAction classes
- switch to G4AnalysisManager

01 September 15: I. Hrivnacova (monopole-V10-01-00)
- Removed EventAction and EventActionMessenger classes, now obsolete, and 
  replaced /testex/event/printModulo commands in macros with 
  /run/printProgress

16 October 14: I. Hrivnacova (monopole-V10-00-00)
- Fixed coding guidelines (long lines) in G4MonopoleTransportation.cc  

03 December 13: I. Hrivnacova (monopole-V09-06-06)
- Updated README for removing AIDA

10 May 13: J.Apostolakis (monopole-V09-06-05)
- G4MonopoleTransportation and G4MonopoleEquation: 
      Adapted to use G4ChargeState to pass Magnetic & Electric Charge, 
          and to changed signature of SetChargeMomentumMass. 

19 February 13: V.Ivantchenko (monopole-V09-06-04)
- G4MonopoleTransportation - removed local variable fVerboseLevel
    and Get/Set methods; use instead verboseLevel variable from 
    the base class

13 February 13: I.Hrivnacova (monopole-V09-06-03)
- Applied coding guidelines 
  (virtual keyword, data members names and initialization)

01 February 13: V.Ivant (monopole-V09-06-02)
- G4MonopolePhysics - fixed warning at initialisation 

31 January 13: V.Ivant (monopole-V09-06-01)
- fixed initialisation after MT was introduced

10 January 13: V.Ivant (monopole-V09-06-00)
- fixed computation of monopole range

20 September 12: I.Hrivnacova (monopole-V09-05-03)
- Adding explicit includes of G4SystemOfUnits.hh and G4PhysicalConstants.hh
  where needed
- Replaced tabulators with space characters  

19 September 12: V.Ivant (monopole-V09-05-02)
- remove not needed file

4 September 12: I.Hrivnacova (monopole-V09-05-01)
- Updated CMakeLists.txt:
  adding visualization, copying macros, install target and comment lines 
- Adding .README  

28 May 12: V.Ivant (monopole-V09-05-00)
- applied code convention

08 December 10: V.Ivant (monopole-V09-03-08)
- added forgotten files

29 November 10: V.Ivant (monopole-V09-03-07)
- Use the most recent version of monopole transportation,
  delta-electron production enabled, non-integer monopole charge allowed

9 November 10: M.Asai (monopole-V09-03-06)
- Fix AIDA file option.

6 June 10: J.Perl (monopole-V09-03-05)
- Remove unused variable in EventAction 

4 June 10: J.Perl (monopole-V09-03-04)
- Updated vis usage 

04 June 10: V.Ivant (monopole-V09-03-03)
- Substituted QGSP by QGSP_BERT
- (J. Apostolakis, B. Bozsogi)  Added new classes for monopole
  G4MonopoleEquation, G4MonopoleFieldSetup, G4MonopoleTransportation

12 May 10: J.Allison (monopole-V09-03-02)
- Introduced GUIExecutive.

31 March 10: V.Ivant (monopole-V09-03-01)
- Removed obsolete header G4MultipleScattering.hh

23 March 10: V.Ivant (monopole-V09-03-00)
- Cleanup and add test on ionisation of heavy highly charged particle

14 July 09: V.Ivant (monopole-V09-02-00)
- Added messenger to define mass and change of the monopole

11 June 08: V.Ivant (monopole-V09-01-00)
- Remove AIDA from GNUmakefile
- Fixed compillation warnings 

16 August 07: V.Ivant (monopole-V09-00-00)
- First release

29 July 22: Pinghan Chu
- Add Mojorana Demonstrator detector structure

src/RunAction.cc
//
// ********************************************************************
// * License and Disclaimer                                           *
// *                                                                  *
// * The  Geant4 software  is  copyright of the Copyright Holders  of *
// * the Geant4 Collaboration.  It is provided  under  the terms  and *
// * conditions of the Geant4 Software License,  included in the file *
// * LICENSE and available at  http://cern.ch/geant4/license .  These *
// * include a list of copyright holders.                             *
// *                                                                  *
// * Neither the authors of this software system, nor their employing *
// * institutes,nor the agencies providing financial support for this *
// * work  make  any representation or  warranty, express or implied, *
// * regarding  this  software system or assume any liability for its *
// * use.  Please see the license in the file  LICENSE  and URL above *
// * for the full disclaimer and the limitation of liability.         *
// *                                                                  *
// * This  code  implementation is the result of  the  scientific and *
// * technical work of the GEANT4 collaboration.                      *
// * By using,  copying,  modifying or  distributing the software (or *
// * any work based  on the software)  you  agree  to acknowledge its *
// * use  in  resulting  scientific  publications,  and indicate your *
// * acceptance of all terms of the Geant4 Software license.          *
// ********************************************************************
//
/// \file exoticphysics/monopole/src/RunAction.cc
/// \brief Implementation of the RunAction class
//
// $Id$
// 
//....oooOO0OOooo........oooOO0OOooo........oooOO0OOooo........oooOO0OOooo......
//....oooOO0OOooo........oooOO0OOooo........oooOO0OOooo........oooOO0OOooo......

#include "RunAction.hh"
#include "RunActionMessenger.hh" 
#include "DetectorConstruction.hh"
#include "PrimaryGeneratorAction.hh"
#include "Run.hh"
#include "G4Event.hh"
#include "G4UnitsTable.hh"
#include "G4SystemOfUnits.hh"
#include "G4ios.hh"
#include "Randomize.hh"

#include "G4EmCalculator.hh"

//....oooOO0OOooo........oooOO0OOooo........oooOO0OOooo........oooOO0OOooo......

RunAction::RunAction(DetectorConstruction* det, PrimaryGeneratorAction* kin)
  :fDetector(det),fKinematic(kin)
{
  fMessenger = new RunActionMessenger(this);
  fBinLength = 5 * CLHEP::mm; 
  G4AnalysisManager* analysisManager = G4AnalysisManager::Instance();  
  analysisManager->SetFileName("monopole");
  analysisManager->SetVerboseLevel(1);
  analysisManager->SetActivation(true);
  
  //G4int eventid = fKinematic->GetEventID();
}

//....oooOO0OOooo........oooOO0OOooo........oooOO0OOooo........oooOO0OOooo......

RunAction::~RunAction()
{
#ifdef G4MULTITHREADED
  if(isMaster) delete fKinematic;
#endif
  delete fMessenger;
}

//....oooOO0OOooo........oooOO0OOooo........oooOO0OOooo........oooOO0OOooo......

G4Run* RunAction::GenerateRun()
{ 
  fRun = new Run(fDetector,fKinematic);
  return fRun;
}

//....oooOO0OOooo........oooOO0OOooo........oooOO0OOooo........oooOO0OOooo......

void RunAction::BeginOfRunAction(const G4Run* aRun)
{
  G4cout << "### Run " << aRun->GetRunID() << " starts." << G4endl;
  //histograms
  //        
  Book();
}

//....oooOO0OOooo........oooOO0OOooo........oooOO0OOooo........oooOO0OOooo......

void RunAction::EndOfRunAction(const G4Run*)
{
  // print Run summary
  //
  if (isMaster) fRun->EndOfRun(fBinLength);    
      
  // save histograms
  G4AnalysisManager* analysisManager = G4AnalysisManager::Instance();  
  if ( analysisManager->IsActive() ) {    
    analysisManager->Write();
    analysisManager->CloseFile();
  }  
}

//....oooOO0OOooo........oooOO0OOooo........oooOO0OOooo........oooOO0OOooo......

void RunAction::SetBinSize(G4double size)
{ 
  fBinLength = size;
  if(fBinLength > fDetector->GetMaxStepSize()) { 
    fBinLength = fDetector->GetMaxStepSize();
  }
}

//....oooOO0OOooo........oooOO0OOooo........oooOO0OOooo........oooOO0OOooo......

void RunAction::Book()
{

  //G4int eventid = anEvent->GetEventID();
  // G4cout << "Event:" << eventid << G4endl;
  G4AnalysisManager* analysisManager = G4AnalysisManager::Instance();  
  analysisManager->SetFirstHistoId(1);   

  G4double length = fDetector->GetAbsorSizeZ();
  G4int nbBins = G4lrint(length / fBinLength);
  // Create tree
  analysisManager->CreateNtuple("MonopoleTree","Monopole Tree");
  //analysisManager->CreateNtupleIColumn("runID"); //0
  //analysisManager->CreateNtupleIColumn("eventID");//1
  analysisManager->CreateNtupleDColumn("PID");//0
  analysisManager->CreateNtupleDColumn("Z");//1
  analysisManager->CreateNtupleDColumn("A");//2
  analysisManager->CreateNtupleDColumn("Charge");//3
  analysisManager->CreateNtupleDColumn("Energy");//4
  analysisManager->CreateNtupleDColumn("EDeposit");//5
  analysisManager->CreateNtupleDColumn("Time");//6
  analysisManager->CreateNtupleDColumn("Weight");//7
  analysisManager->CreateNtupleDColumn("PosX");//8
  analysisManager->CreateNtupleDColumn("PosY");//9
  analysisManager->CreateNtupleDColumn("PosZ");//10
  analysisManager->CreateNtupleIColumn("TrackID");//11
  analysisManager->CreateNtupleIColumn("ParentID");//12
  analysisManager->CreateNtupleDColumn("Velocity");//13
  analysisManager->CreateNtupleDColumn("U");//14
  analysisManager->CreateNtupleDColumn("V");//15
  analysisManager->CreateNtupleDColumn("W");//16
  analysisManager->CreateNtupleIColumn("EventID");//17
  analysisManager->CreateNtupleSColumn("VolumeName");//18
  //analysisManager->CreateNtupleIColumn("Channel");//19
  analysisManager->FinishNtuple();
  // Create histograms
  analysisManager->CreateH1("h1","Edep (MeV/mm) along absorber (mm)", 
                            nbBins, 0, length);
  analysisManager->CreateH1("h2","DEDX (MeV/mm) of proton", 100, -3., 7.);
  analysisManager->CreateH1("h3","DEDX (MeV/mm) of monopole", 100, -3., 7.);
  analysisManager->CreateH1("h4","Range(mm) of proton", 100, -3., 7., "mm");
  analysisManager->CreateH1("h5","Range(mm) of monopole", 100, -3., 7., "mm");
  analysisManager->OpenFile(); 
}

//....oooOO0OOooo........oooOO0OOooo........oooOO0OOooo........oooOO0OOooo......
